crates/turborepo-lib/src/process/child.rs

```diff
@@ -77,27 +83,242 @@ impl From<std::io::Error> for ShutdownFailed {
+        let mut child = command.spawn()?;
+        let pid = child.id();
+
+        let stdin = child.stdin.take().map(ChildInput::from);
+        let stdout = child.stdout.take().map(ChildOutput::Concrete);
+        let stderr = child.stderr.take().map(ChildOutput::Concrete);
+
+        Ok((
+            Self {
+                pid,
+                imp: ChildHandleImpl::Tokio(child),
+            },
+            ChildIO {
+                controller: None,
+                stdin,
+                stdout,
+                stderr,
+            },
+        ))
+    }
+
+    pub fn spawn_pty(command: Command) -> io::Result<(Self, ChildIO)> {
+        use portable_pty::PtySize;
+
+        let keep_stdin_open = command.will_open_stdin();
+
+        let command = portable_pty::CommandBuilder::from(command);
+        let pty_system = native_pty_system();
+        // TODO we should forward the correct size
+        let pair = pty_system
+            .openpty(PtySize::default())
+            .map_err(|err| match err.downcast() {
+                Ok(err) => err,
+                Err(err) => io::Error::new(io::ErrorKind::Other, err),
+            })?;
+
+        let controller = pair.master;
+        let receiver = pair.slave;
+
+        let child = receiver
+            .spawn_command(command)
+            .map_err(|err| match err.downcast() {
+                Ok(err) => err,
+                Err(err) => io::Error::new(io::ErrorKind::Other, err),
+            })?;
+
+        let pid = child.process_id();
+
+        let mut stdin = controller.take_writer().ok().map(ChildInput::from);
+        let stdout = controller.try_clone_reader().ok().map(ChildStdout::from);
+
+        // If we don't want to keep stdin open we take it here and it is immediately
+        // dropped resulting in a EOF being sent to the child process.
+        if !keep_stdin_open {
+            stdin.take();
+        }
+
+        Ok((
+            Self {
+                pid,
+                imp: ChildHandleImpl::Pty { receiver, child },
+            },
+            ChildIO {
+                controller: Some(controller),
+                stdin,
+                stdout,
+                stderr: None,
+            },
+        ))
+    }
+
+    pub fn pid(&self) -> Option<u32> {
+        self.pid
+    }
+
+    pub async fn wait(&mut self) -> io::Result<Option<i32>> {
+        match &mut self.imp {
+            ChildHandleImpl::Tokio(child) => child.wait().await.map(|status| status.code()),
+            ChildHandleImpl::Pty { child, .. } => {
+                // TODO: we currently poll the child to see if it has finished yet which is less
+                // than ideal
+                loop {
+                    match child.try_wait() {
+                        // This is safe as the portable_pty::ExitStatus's exit code is just
+                        // converted from a i32 to an u32 before we get it
+                        Ok(Some(status)) => return Ok(Some(status.exit_code() as i32)),
```

Comment:
At the moment, `portable_pty` converts exit statuses caused by signals to have an exit code `1` and it captures the signal that the process received. At the moment there's no way to extract this signal aside from displaying the `portable_pty::ExitStatus`. I'll put an upstream PR so we can differentiate between non-zero exits and signals.